services/supervisord/pmm_update_checker.go

```diff
@@ -108,41 +105,50 @@ func (p *PMMUpdateChecker) Installed() *version.PackageInfo {
 		return nil
 	}
 
-	p.rw.Lock()
+	p.installedRW.Lock()
 	p.lastInstalledPackageInfo = &res.Installed
-	p.rw.Unlock()
+	p.installedRW.Unlock()
 
 	return &res.Installed
 }
 
+func (p *PMMUpdateChecker) cmdRun(ctx context.Context, cmdLine string) ([]byte, bytes.Buffer, error) {
+	args := strings.Split(cmdLine, " ")
+	p.cmdMutex.Lock()
+	timeoutCtx, cancel := context.WithTimeout(ctx, updateDefaultTimeout)
+	defer cancel()
+	cmd := exec.CommandContext(timeoutCtx, args[0], args[1:]...) //nolint:gosec
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	pdeathsig.Set(cmd, unix.SIGKILL)
+
+	b, err := cmd.Output()
+	p.cmdMutex.Unlock()
+	return b, stderr, err
+}
+
 // checkResult returns last `pmm-update -check` result and check time.
 // It may force re-check if last result is empty or too old.
-func (p *PMMUpdateChecker) checkResult() (*version.UpdateCheckResult, time.Time) {
-	p.rw.RLock()
-	defer p.rw.RUnlock()
+func (p *PMMUpdateChecker) checkResult(ctx context.Context) (*version.UpdateCheckResult, time.Time) {
+	p.checkRW.RLock()
+	defer p.checkRW.RUnlock()
 
 	if time.Since(p.lastCheckTime) > updateCheckResultFresh {
```

Comment:
Can we make this atomic, by creating another function to get `lastCheckTime` (or some other way) instead of having a defer unlock above and an unlock/lock pair inside the if?

Reply:
this lock not only for `p.lastCheckTime`, but for `p.lastCheckResult` too. We have to unlock Read to be able to lock Write in `check`